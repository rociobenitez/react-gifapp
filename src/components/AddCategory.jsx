import { useState } from 'react'
import PropTypes from 'prop-types'
import { Input } from '@mui/material'

export const AddCategory = ({ onNewCategory }) => {

    const [ inputValue, setInputValue ] = useState('')

    const onChange = ({ target }) => {     // event.target
        setInputValue( target.value )      // event.target.value
    }

    const onSubmit = ( ev ) => {
        ev.preventDefault()
        const value = inputValue.trim()
        if( value.length <= 1 ) return    // validation

        setInputValue('')                 // clear text box
        onNewCategory( value )
    }

    return (
        <form onSubmit = { onSubmit } aria-label='form'>        {/* (ev) => onSubmit(ev) */}
            <Input
                fullWidth 
                type        = "text"
                placeholder = "Buscar gifs"
                value       = { inputValue }
                onChange    = { onChange }  
                sx          = {{ margin: '20px 0', padding:'10px' }}
            />
        </form>   
    )
}


AddCategory.propTypes = {
    onNewCategory : PropTypes.func.isRequired
}