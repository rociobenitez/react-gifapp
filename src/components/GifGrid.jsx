import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';
import { ImageItem } from '.';

export const GifGrid = ({ category }) => {

    const { images, isLoading } = useFetchGifs( category )

    return (
        <>  
            {
                isLoading && ( <h3>Cargando...</h3> )
            }
            {
                images.map( image => (
                    <ImageItem
                        key = { image.id }
                        { ...image }
                    />
                ))
            }
        </>
    )
}

GifGrid.propTypes = {
    category : PropTypes.string.isRequired
}