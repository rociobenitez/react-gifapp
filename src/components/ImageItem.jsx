import { ImageListItem } from "@mui/material"
import PropTypes from 'prop-types'

export const ImageItem = ({ id, title, url }) => {
    return (
        <ImageListItem key = { id }>
            <img
                src      = { `${url}?w=164&h=164&fit=crop&auto=format` }
                srcSet   = { `${url}?w=164&h=164&fit=crop&auto=format&dpr=2 2x` }
                alt      = { title }
                loading  = "lazy"
            />
        </ImageListItem>
    )
}

ImageItem.propTypes = {
    id    : PropTypes.number,
    title : PropTypes.string.isRequired,
    url   : PropTypes.string.isRequired,
}