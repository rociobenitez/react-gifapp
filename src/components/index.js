export * from './AddCategory'
export * from './GifGrid'
export * from './Header'
export * from './ImageItem'
