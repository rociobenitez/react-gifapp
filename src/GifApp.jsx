import { useState } from "react"
import { AddCategory, GifGrid, Header } from "./components"
import { ImageList } from "@mui/material"

export const GifApp = () => {
    
    const [ categories, setCategories ] = useState([])

    const onAddCategory = ( newCategory ) => {
        if( categories.includes( newCategory ) ) return      // avoid repeating categories
        setCategories( [ newCategory, ...categories ] )
    }

    return (
        <>
            <Header />
            
            <AddCategory
                onNewCategory = { onAddCategory }    // { (value) => onAddCategory(value) }
            />

            <ImageList variant="masonry" cols={4} gap={8}>
                {
                    categories.map( (category) => (
                        <GifGrid
                            key      = { category }
                            category = { category }
                        />
                    ))
                }               
            </ImageList>


        </>
    )
}