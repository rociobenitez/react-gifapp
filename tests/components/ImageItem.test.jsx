import { render, screen } from "@testing-library/react"
import { ImageItem } from "../../src/components"

describe('Tests in <ImageItem />', () => { 
    
    const id    = 1
    const title = 'cheese bar'
    const url   = 'https://gifapp.com/cheese-bar'

    test('match with the snapshot', () => {
        const { container } = render(<ImageItem id = { id } title = { title } url = { url } />)
        expect( container ).toMatchSnapshot()
    })

    test('should show the image with the url', () => {
        render(<ImageItem id = { id } title = { title } url = { url } />)
        const { src, alt } = screen.getByRole('img')
        expect( src ).toBe( url + '?w=164&h=164&fit=crop&auto=format' )
        expect( alt ).toBe( title )
    })
})