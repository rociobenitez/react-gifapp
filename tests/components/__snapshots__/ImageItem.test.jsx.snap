// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <ImageItem /> match with the snapshot 1`] = `
<div>
  <li
    class="MuiImageListItem-root css-l086p4-MuiImageListItem-root"
    style="height: auto; grid-column-end: span 1; grid-row-end: span 1;"
  >
    <img
      alt="cheese bar"
      class="MuiImageListItem-img"
      loading="lazy"
      src="https://gifapp.com/cheese-bar?w=164&h=164&fit=crop&auto=format"
      srcset="https://gifapp.com/cheese-bar?w=164&h=164&fit=crop&auto=format&dpr=2 2x"
    />
  </li>
</div>
`;
